src/Umbraco.Core/Services/UserGroupService.cs

```diff
@@ -0,0 +1,308 @@
+            .ToArray();
+
+        return Task.FromResult(contents);
+    }
+
+    /// <inheritdoc />
+    public Task<IUserGroup?> GetAsync(string alias)
+    {
+        if (string.IsNullOrWhiteSpace(alias))
+        {
+            throw new ArgumentException("Value cannot be null or whitespace.", nameof(alias));
+        }
+
+        using ICoreScope scope = ScopeProvider.CreateCoreScope(autoComplete: true);
+
+        IQuery<IUserGroup> query = Query<IUserGroup>().Where(x => x.Alias == alias);
+        IUserGroup? contents = _userGroupRepository.Get(query).FirstOrDefault();
+        return Task.FromResult(contents);
+    }
+
+    /// <inheritdoc />
+    public Task<IUserGroup?> GetAsync(int id)
+    {
+        using ICoreScope scope = ScopeProvider.CreateCoreScope(autoComplete: true);
+        return Task.FromResult(_userGroupRepository.Get(id));
+    }
+
+    /// <inheritdoc />
+    public Task<IUserGroup?> GetAsync(Guid key)
+    {
+        using ICoreScope scope = ScopeProvider.CreateCoreScope(autoComplete: true);
+
+        IQuery<IUserGroup> query = Query<IUserGroup>().Where(x => x.Key == key);
+        IUserGroup? groups = _userGroupRepository.Get(query).FirstOrDefault();
+        return Task.FromResult(groups);
+    }
+
+    /// <inheritdoc/>
+    public async Task<Attempt<UserGroupOperationStatus>> DeleteAsync(Guid key)
+    {
+        IUserGroup? userGroup = await GetAsync(key);
+
+        Attempt<UserGroupOperationStatus> validationResult = ValidateUserGroupDeletion(userGroup);
+        if (validationResult.Success is false)
+        {
+            return validationResult;
+        }
+
+        EventMessages eventMessages = EventMessagesFactory.Get();
+
+        using (ICoreScope scope = ScopeProvider.CreateCoreScope())
+        {
+            var deletingNotification = new UserGroupDeletingNotification(userGroup!, eventMessages);
+
+            if (await scope.Notifications.PublishCancelableAsync(deletingNotification))
+            {
+                scope.Complete();
+                return Attempt.Fail(UserGroupOperationStatus.CancelledByNotification);
+            }
+
+            _userGroupRepository.Delete(userGroup!);
+
+            scope.Notifications.Publish(new UserGroupDeletedNotification(userGroup!, eventMessages).WithStateFrom(deletingNotification));
+
+            scope.Complete();
+        }
+
+        return Attempt.Succeed(UserGroupOperationStatus.Success);
+    }
+
+    private Attempt<UserGroupOperationStatus> ValidateUserGroupDeletion(IUserGroup? userGroup)
+    {
+        if (userGroup is null)
+        {
+            return Attempt.Fail(UserGroupOperationStatus.NotFound);
+        }
+
+        if (userGroup.IsSystemUserGroup())
+        {
+            return Attempt.Fail(UserGroupOperationStatus.IsSystemUserGroup);
+        }
+
+        return Attempt.Succeed(UserGroupOperationStatus.Success);
+    }
+
+    /// <inheritdoc />
+    public async Task<Attempt<IUserGroup, UserGroupOperationStatus>> CreateAsync(
```

Comment:
Very good point, I've updated the ViewModelFactory to also check, as well as the service. 

I also had to implement a bit extra on our `EntityService` we already had an `Exists` method, however I've added one where you can specify the object type that should be existing, otherwise you'd be able to specify a document key as a media start node key 😄